config/application.rb

```diff
@@ -11,7 +11,7 @@ class Application < Rails::Application
 
     # allow nested structure in Models directory without additional namespacing
     # from http://stackoverflow.com/questions/18934115/rails-4-organize-rails-models-in-sub-path-without-namespacing-models
-    #config.autoload_paths += Dir[Rails.root.join('app', 'models', '{**}')]
+    config.autoload_paths << Rails.root.join('lib')
```

Comment:
@Tuuleh I agree with you on the importance of namespacing, and it's something that ruby does so well. But Rails' autoloading doesn't prohibit us from nesting models within the `app` dir. Tell me if I've misunderstood your comment.

@SumOfUs/developers  FYI - rails' default autoloading within the `app` directory is rather good, so long as certain rules are obeyed. Basically, have your class/module name match the name of the file. And if a file is within a sub directory, make sure the class/module is correctly namespaced:

```ruby

# GOOD
#
# app/models/foo.rb
class Foo
end

# app/models/foo/bar.rb
class Foo::Bar
end

# app/models/nested/foo_bar.rb
module Nested
  class FooBar
  end
end

# BAD
#
# app/models/foo.rb
class Bar::Foo
end

class NotFoo
end

# app/models/foo/bar.rb
class Bar
end
```

From experience, I've found trying to fight these conventions by appending patterns to  `config.autoload_paths` for `/app` can be really problematic down the road (files not loading when they should and strange name conflicts).